File: 2025_06/leetcode2025-06-26.cpp
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    int longestSubsequence(string s, int k) {
        int n = s.size();
        long long cnt = 0, v = 0;
        for (int i=n-1; i>=0; i--){
            if (s[i] == '0'){
                cnt++;
            }
            else{
                if (cnt<30){
                    // 这里是为了防止左移时超出longlong限制
                    // 而30是因为题目中最大数字1e9的二进制位数正好是2+30位，前面是0b需要占两位，所以正好是30位数字
                    v |= (1ll<<cnt);
                    if (v<=k){
                        cnt++;
                    }
                }
            }
        }
        return cnt;
    }
};

int main(){
    Solution sl;
    string s = "1001010";
    int k = 5;
    cout << sl.longestSubsequence(s, k) << endl;
}

File: 2025_06/leetcode2025-06-26.py
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        # 从右往左遍历，只要累加的二进制对应的十进制数字还没有超过k就可以继续在前面累加1，而如果超过了k，我们就只能计算前面0的个数而不能计算1的个数了
        # 在程序中我们用v来维护当前已经记录的二进制数字,并且用cnt来维护v的二进制位数,那么每一次在v之前加一个1时就直接用v或上1<<cnt即可实现
        cnt = 0
        v = 0
        n = len(s)
        for i in range(n-1, -1, -1):
            if s[i] == '0':
                cnt += 1
            elif (v|(1<<cnt))<=k:
                v |= (1<<cnt)
                cnt += 1
        return cnt

if __name__ == "__main__":
    sl = Solution()
    s = "1001010"
    k = 5
    print(sl.longestSubsequence(s, k))
